Add Database.files_duration()

frankenjoe opened the pull request:
As discussed in #111 this adds `Database.file_duration()`.

![image](https://user-images.githubusercontent.com/10383417/140804334-10b3766b-190e-4092-bb8e-697a3c6fe2f3.png)

Uses `audiofile.duration()` to calculate the duration of a file and caches the result Internally in `Database._files_duration`. I decided to name it `files_duration()` to not confuse it with `utils.duration()` which returns the duration of the segments in a table.

I decided to use a dictionary instead of a `pd.Series`, as the latter is much slower as the following test shows:

```python
ts = [pd.to_timedelta(random.random(), unit='s') for _ in range(10000)]

d = {}
tt = time.time()
for idx, t in enumerate(ts):
    d[f'{idx}'] = t
time.time() - tt
```
```
0.0021936893463134766
```

```python
y = pd.Series()
tt = time.time()
for idx, t in enumerate(ts):
    y[f'{idx}'] = t
time.time() - tt
```
```
9.314390182495117
```

### Example

```python
db = audb.load(
    'emodb',
    version='1.1.1',
    format='wav',
    mixdown=True,
    sampling_rate=16000,
    only_metadata=True,
    full_path=False,
)
```

```python
db.files_duration(db.files).sum()
```
```
0 days 00:24:47.092187490
```

```python
db.files_duration(db['emotion'].files[:10])
```
```
file
wav/03a01Fa.wav      0 days 00:00:01.898250
wav/03a01Nc.wav      0 days 00:00:01.611250
wav/03a01Wa.wav   0 days 00:00:01.877812500
wav/03a02Fc.wav      0 days 00:00:02.006250
wav/03a02Nc.wav   0 days 00:00:01.439812500
wav/03a02Ta.wav   0 days 00:00:01.735687500
wav/03a02Wb.wav      0 days 00:00:02.123625
wav/03a02Wc.wav   0 days 00:00:01.498062500
wav/03a04Ad.wav   0 days 00:00:01.504874999
wav/03a04Fd.wav   0 days 00:00:01.696812500
Name: file, dtype: timedelta64[ns]
```

hagenw commented on audformat/core/database.py:
```diff
@@ -289,6 +291,40 @@ def drop_tables(
+                path = file
+            dur = audiofile.duration(path)
+            dur = pd.to_timedelta(dur, unit='s')
+            self._file_duration[file] = dur
```
Above you allow that `file` is an absolute path, which is ok for running `audiofile` on it. But when we then store it allowing also relative or absolute path, it means that it `self._file_duration` can get twice as big as `db.files` in the end (or even bigger if the user moves around the database in between).

hagenw replied:
One way around this would be that we allow only the syntax that is used in the current index (e.g. `db.files`).
But again, `db.files` is generated on the fly and it will change if a user updates a table in between.

frankenjoe replied:
I guess whether `file` is absolute or relative depends mainly on how the files are stored in the table. We could of course always make it relative by removing `db.root`, but I'm not sure if this really worth the effort. In the tables we also usually store absolute paths.

frankenjoe replied:
I also don't think moving around a database is a common use case, is it?

frankenjoe replied:
On the other hand - always storing the relative path would indeed solve the problem :)

hagenw replied:
To not have to deal with https://github.com/audeering/audformat/pull/114#discussion_r744586885 maybe we indeed only store the relative path of the files. 

> I also don't think moving around a database is a common use case, is it?

It is if you do `db.update(db_new)`, where `db_new` is moved to the new location. Will it in general work if both `db` and `db_new` contain already entries in `db._files_duration`?

frankenjoe replied:
But `db.update()` just add some new files, I don't see why this is a problem.

hagenw replied:
The question was more what will happen with its `_files_duration` entries. Will they be merged with the ones from `db`? If yes, it will only make sense if the path is stored relative as the absolute path would now be different.

frankenjoe replied:
Mhh, yes if we store the file paths relative we could indeed update `_files_duration`. Otherwise, it would simply re-calculate the duration for the new files when `files_duration()` is called.

At the moment I'm not completely convinced if we should really store the relative path, though. Cause then there is of course the question what `files_duration()` should return. Always the relative path? Always the full path? At the moment it will return it the same way it's stored in the tables. And I don't think that we will often have the case that the user requests a database with absolute path, but then calls `files_duration()` or vice versa. At least as long as the user passes `db.files` or `db[table_id].files` this is not the case. And even if the user does it, the result will still be correct, though it has re-calculate the duration in that case. But as I said, I don't see it as a common use case.

hagenw replied:
I'm also fine with your proposal. To summarize it:

* `db.update(db_new)` will simply ignore the entries in `db_new._files_duration`
* in `db._files_duration` we store relative and absolute paths as different entries. This is not a problem as the user should not access this dictionary
* we could simply recalculate the duration if a user has called first absolute path and then relative (or vice versa), but I think we could even avoid it, see https://github.com/audeering/audformat/pull/114#discussion_r744586885

hagenw replied:
If I see it correctly the first two points are already implemented by the current behavior, correct?

frankenjoe replied:
Yes, and with the latest commit the third point should be implemented, too. See my comment at #114.

hagenw commented on tests/test_database.py:
```diff
@@ -163,6 +164,28 @@ def test_drop_and_pick_tables():
+    full_file = os.path.join(db.root, file)
+    dur = pd.to_timedelta(audiofile.duration(full_file), unit='s')
+    assert db.file_duration(file) == dur
+    assert db.file_duration(full_file) == dur
```
I think you should repeat the last two lines in order to check that a second call still returns the same duration.

frankenjoe replied:
done

hagenw commented:
I agree that it might be dangerous to name it `Database.duration()`. The problem with the current solution of `Database.file_duration()` is that I would have never guessed that I can use it to get the overall duration. Instead I would have tried the following:
```python
duration = np.sum([db.file_duration(file) for file in db.files])
```
I don't have an obvious solution for the problem. We could use `Database.files_duration()`, but this does also not look very nice.

frankenjoe commented:
> The problem with the current solution of `Database.file_duration()` is that I would have never guessed that I can use it to get the overall duration.

We could also make `file` a mandatory argument and add a property `Database.total_file_duration` instead.

hagenw commented:
How do you solve the problem of an updated file, that has already a duration entry?
I guess at the moment the idea is that this is only correctly handled in `audb`, but if a user modifies a local database we don't check for this.

frankenjoe commented:
>  The only alternative to make it safe inside `audformat` would be to always call `audiofile` as the underlying file could always be changed/updated by the user

Yes, but then we don't gain anything. The ultimate goal is that we make the conversion from filewise to segmented index without `NaT` faster.

> We could of cause also argue that this is not the way you should use audformat and simply document it?

Yes, if a user changes a file under the hood, the labels will also no longer fit. So this is of course a general problem.

frankenjoe commented:
> We could use `Database.files_duration()`, but this does also not look very nice.

I now renamed the function to `Database.files_duration()`. It now also accepts a sequence of files and in that case returns a mapping from file to duration. A user can now use `db.files_duration(db.files).sum()` to get the total duration of a database.

frankenjoe commented:
I updated the description and added an example there. Would be nice to add those to the docstring directly. So maybe we should think about loading a database with `audb` when we build the documentation.

hagenw commented:
> The ultimate goal is that we make the conversion from filewise to segmented index without setting end=NaT faster.

How can we achieve this for `audformat.utils.to_segmented_index()` as this has no access to the `Database` object?
The same problem we phase for `audformat.utils.duration()`.

frankenjoe commented:
> How can we achieve this for `audformat.utils.to_segmented_index()` as this has no access to the `Database` object?
> The same problem we phase for `audformat.utils.duration()`.

Yes, that's why I suggested to add a `as_segmented` option to `Table.get()` and `Column.get()`. I.e. the user can replace:

```python
df = db['table'].get()
df = utils.to_segmented_index(df, allow_nat=False)
```

with:

```
df = db['table'].get(as_segmented=True, allow_nat=False)
```

hagenw commented on audformat/core/database.py:
```diff
@@ -289,6 +290,47 @@ def drop_tables(
+        Use ``db.files_duration(db.files).sum()``
+        to get the total duration of a database.
+        Or ``db.files_duration(db['table'].files).sum()``
+        to get the total duration of a table.
```
It's not so nice that we now have two functions that can do the same thing. E.g. `db.files_duration(db.files).sum()`
and `audformat.utils.duration(db.files)`. As I don't have to add the `.sum()` at the end, I think I would pick the later as a user as I will not be aware that it cannot use the caching.

frankenjoe replied:
The advantage I see is that you can easily apply other operations, e.g. `db.files_duration(db.files).max()` or  `db.files_duration(db.files).mean()`.

hagenw replied:
Yes, I also see the advantage. I'm just worried that most of the users will simply continue using `audformat.utils.duration()`, but I think we simply have to life with it. It's also still great to have it as it provides the possibility to calculate duration for data frames that are not part of any database.

hagenw replied:
I think it is not obvious what "the total duration of a table" means for a segmented table.
Maybe we should change to
```suggestion
        to get the total duration of all files assigned to a table.
```

frankenjoe replied:
Changed to:

![image](https://user-images.githubusercontent.com/10383417/140722180-63ee090f-779a-4776-a9cd-4408937bbf5c.png)

hagenw commented on audformat/core/database.py:
```diff
@@ -289,6 +290,47 @@ def drop_tables(
+    def files_duration(
+            self,
+            files: typing.Union[str, typing.Sequence[str]],
+    ) -> typing.Union[pd.Timedelta, pd.Series]:
```
I think it is not a good idea to return different things for a single file or several files. I would change to always return a series.

frankenjoe replied:
Changed it accordingly.

hagenw commented:
> > How can we achieve this for `audformat.utils.to_segmented_index()` as this has no access to the `Database` object?
> > The same problem we phase for `audformat.utils.duration()`.
> 
> Yes, that's why I suggested to add a `as_segmented` option to `Table.get()` and `Column.get()`. I.e. the user can replace:
> 
> ```python
> df = db['table'].get()
> df = utils.to_segmented_index(df, allow_nat=False)
> ```
> 
> with:
> 
> ```python
> df = db['table'].get(as_segmented=True, allow_nat=False)
> ```

Ah, yes, found it at https://github.com/audeering/audformat/issues/111#issue-1043261735 thanks.
This seems to be a possible solution.

We then just need to train the users when to use `audformat.utils.to_segmented_index()` and when not ;)

hagenw commented on audformat/core/database.py:
```diff
@@ -289,6 +290,47 @@ def drop_tables(
+
+        Use ``db.files_duration(db.files).sum()``
+        to get the total duration of a database.
+        Or ``db.files_duration(db['table'].files).sum()``
```
I would change it to
```suggestion
        Or ``db.files_duration(db[table_id].files).sum()``
```

frankenjoe replied:
Changed it accordingly.

hagenw commented:
The example you provide above fails for me, because due to `only_metadata=True` there are no files:

```python
db = audb.load(
    'emodb',
    version='1.1.1',
    format='wav',
    mixdown=True,
    sampling_rate=16000,
    only_metadata=True,
    full_path=False,
)
db.files_duration(db['emotion'].files[:10])
```
results in
```
In [9]: db.files_duration(db['emotion'].files[:10])                                        
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-a1e8b2b5ac20> in <module>
----> 1 db.files_duration(db['emotion'].files[:10])
                                                  
~/git/audeering/audformat/audformat/core/database.py in files_duration(self, files)
    327                                                                                                                                         
    328         files = audeer.to_list(files)                                     
--> 329         return pd.Series(files, index=files).map(duration)       
    330                                                         
    331     def map_files(                                        
                                                                            
~/.envs/audformat/lib/python3.6/site-packages/pandas/core/series.py in map(self, arg, na_action)
   3981         dtype: object                                                                     
   3982         """                                                         
-> 3983         new_values = super()._map_values(arg, na_action=na_action)
   3984         return self._constructor(new_values, index=self.index).__finalize__(
   3985             self, method="map"        
                                                                               
~/.envs/audformat/lib/python3.6/site-packages/pandas/core/base.py in _map_values(self, mapper, na_action)
   1158                                                                                
   1159         # mapper is a function
-> 1160         new_values = map_f(values, mapper) 
   1161                                                                                     
   1162         return new_values
       
pandas/_libs/lib.pyx in pandas._libs.lib.map_infer()
                                                                                                                       
~/git/audeering/audformat/audformat/core/database.py in duration(file)
    321                 else:
    322                     path = file
--> 323                 dur = audiofile.duration(path)
    324                 dur = pd.to_timedelta(dur, unit='s')
    325                 self._files_duration[file] = dur
                                 
...

RuntimeError: Error opening '/home/audeering.local/hwierstorf/audb/emodb/1.1.1/fe182b91/wav/03a01Nc.wav': System error.
```

I guess we need to handle this case. The question is if we should return an empty Series then, or a Series containing all the files, but no durations or 0 as duration?

hagenw commented on audformat/core/database.py:
```diff
@@ -289,6 +290,44 @@ def drop_tables(
+                    path = os.path.join(self.root, file)
+                else:
+                    path = file
+                dur = audiofile.duration(path)
```
I think we should add another check here.

Let us assume that all paths are stored absolute in `db._files_duration` because the database was loaded with `full_path=True` and then we called `db.files_duration(db.files)`. If I then call `db.files_duration(relative_path_to_file)` it will generate the correct path with
```python
path = os.path.join(self.root, file)
```
but after that we should not directly call `audiofile.duration(path)` but could check again if `path` is inside `self._files_duration` and use that value instead.

frankenjoe replied:
I added a check as you suggested. In addition I applied two more changes:

1. I set the name of the returned `pd.Series` to `file`
2. I added an argument `root` to cover the case where file names are relative but the database was not saved yet or loaded from disk (e.g. when a user prepares a database for publishing). In this case the user has to provide `root` or otherwise an error is raised.

hagenw replied:
Great, thanks.

frankenjoe commented:
> I guess we need to handle this case. The question is if we should return an empty Series then, or a Series containing all the files, but no durations or 0 as duration?

Doesn't it make sense to raise an error if files do not exist? If we silently ignore it, the user may not notice it and work with wrong results.

hagenw commented:
> > I guess we need to handle this case. The question is if we should return an empty Series then, or a Series containing all the files, but no durations or 0 as duration?
> 
> Doesn't it make sense to raise an error if files do not exist? If we silently ignore it, the user may not notice it and work with wrong results.

It would be ok to raise an error, but it will have consequences. E.g. we should then not set `allow_nat=False` in `get()`, because users will work a lot with `only_metadata=True` and still might want to convert to a segmented table. BTW, for some tables it will even work if only the metadata is downloaded as `audb` will fill in the `db._files_duration` table, but this will not be the case for all databases.

frankenjoe commented:
> E.g. we should then not set `allow_nat=False` as default in `get()`

Yes, in `utils.to_segmented_table()` we also have `allow_nat=True` as the default value.

frankenjoe commented:
>  BTW, for some tables it will even work if only the metadata is downloaded as `audb` will fill in the `db._files_duration` table, but this will not be the case for all databases.

True, for all tables loaded with `audb` it will work even if the files were not downloaded!

frankenjoe commented:
> No, not for all, we still have databases published that don't have duration stored in the dependency table:

Ah yeah, right. I guess it applies to databases we have directly converted from the old `audb` repository. But for databases published with `audb.publish()` it should work, correct?

hagenw commented:
> > No, not for all, we still have databases published that don't have duration stored in the dependency table:
> 
> Ah yeah, right. I guess it applies to databases we have directly converted from the old `audb` repository. But for databases published with `audb.publish()` it should work, correct?

Yes, at some point we decided to switch to use `audiofile.duration(file, sloppy=True)` and that the duration should also be stored for MP3 and MP4 files when running `audb.publish()`. At the beginning we set duration to `0.0` for MP3 and MP4 files. But in general, I don't think this is something we have to worry about here.

hagenw commented on audformat/core/database.py:
```diff
@@ -289,6 +290,82 @@ def drop_tables(
+            files: file names
+            root: root directory under which the files are stored.
+                Provide if file names are relative and
+                database was not saved or loaded from disk
```
Maybe we should add here, what the default behavior is, e.g.

```suggestion
                database was not saved or loaded from disk.
                If ``None`` :attr:`audformat.Database.root` is used
```

frankenjoe replied:
Good idea.
